### DISCUSSION TITLE
Possible memory leak in Skia.Gtk heap

### DISCUSSION THREAD

**Discussion by mcNets** [unoplatform/uno#12978] [Q&A]
Hi.

I developed an app that runs on a Linux device controlling a machine and I have noticed that is increasingly consuming a lot of memory.

I am trying to detect the memory leak using the Performance profiler tool, I am sorry but I am not used to it. What I did is:

1. Run the Skia.Gtk heap in my computer and take some snapshots.
2. Run the Windows heap and take again the aprox same  snapshots.

And these are the results:

**Windows**
![image](https://github.com/unoplatform/uno/assets/24267381/b0ddb623-7b45-4321-b5ff-184172fb47a8)

**Skia.Gtk**
![image](https://github.com/unoplatform/uno/assets/24267381/028d4cea-9b15-43bd-a8c8-d5be865ebc4b)

Any hint on how can I dig deeper and find the memory leak?

The app is very simple, just send gRPC messages to a service and receive broadcast notifications from it.

[Report20230722-1430.zip](https://github.com/unoplatform/uno/files/12136877/Report20230722-1430.zip)

**Comment by jeromelaban** [comment]
Thanks for the report. The trace you've provided is done on the `.Windows` head, not on the `skia.gtk` head. If there's a leak in the `.Windows` head, it's not something Uno can adjust and you'll want to report it in the official winui repository.

**Reply by mcNets** [reply]
Hi @jeromelaban.
Maybe I sent one report instead of the other. I'll try to build it again and I'll also try to build another project that reproduces the memory leak.

**Reply by mcNets** [reply]
@jeromelaban I attached a minimum example that reproduces the possible memory leak.

I'll thank you If you could take a look.

Run the 'testService' project on a Windows Terminal and either testUNO.Skia or testUNO.Windows in Visual Studio. testUno.Skia consumes more memory.

I've used the Performance profiler - Memory usage but I'm not able to see why the memory increases.

[testMem.zip](https://github.com/unoplatform/uno/files/12175500/testMem.zip)

**Reply by jeromelaban** [reply]
Thanks. Could you still provide a memory trace like you did, but for the skia app?

**Reply by mcNets** [reply]
I'll send you today. Thanks.

**Reply by mcNets** [reply]
[sysmacui.skia.gtk.zip](https://github.com/unoplatform/uno/files/12610623/sysmacui.skia.gtk.zip)

**Reply by jeromelaban** [reply]
Thanks for the update. What's leaking is unclear, but is having 24 `SysmacUI.Controls.KPIControl` instances something that should be happening? Also, there are 5 instances of `MainMenuPage`, which could indicate multiple forward navigation? 

You may want to try to simplify your pages to see what is leaking.

**Reply by mcNets** [reply]
@jeromelaban thanks a lot for your time and help.

I don't understand why there are 5 instances of ``MainMenuPage` 

This is how my navigation service looks like, it doesn't allow to navigate to the current page:

```#csharp
internal class NavService {

    public static void NavigateTo(string pageName, object? parameters) {
        var app = (App)Application.Current;
        var window = app.MainWindow;

        if (window == null || window.Content == null) {
            Debug.WriteLine("NavigateTo requires a Window.Content");
            return;
        }

        string aqn = $"SysmacUI.Pages.{pageName}";
        Type? type = Type.GetType(aqn);
        if (type == null) {
            Debug.WriteLine($"There is no {aqn}");
            return;
        }

        var frame = (Frame)window.Content;
        if (frame.Content.GetType() == type) {
            Debug.WriteLine($"Already in {aqn}");
            return;
        }

        frame?.Navigate(type, parameters);
    }

    public static void MainMenu() => NavigateTo("MainMenuPage", null);
    public static void Autocontrol() => NavigateTo("AutocontrolPage", null);
    public static void Controls() => NavigateTo("ControlsPage", null);

    public static async void GPIO() {
        if (await DlgService.ShowRfidDlgAsync() == false) {
            return;
        }
        NavigateTo("GPIOPage", null);
    }
```

And I use it in XAML in this way:

```
<controls:BigButtonControl Grid.Row="0"
                           Grid.Column="3"
                           Text="PRODUCCIO"
                           Background="{StaticResource SysmacGradient4}"
                           ButtonWidth="150"
                           HorizontalAlignment="Center"
                           Margin="0,0,0,5"
                           OnClick="{x:Bind serv:NavService.Production}" />
```

Let me start again with a simple project and see what happens.

I don't think my company will accept to pay for technical support, and I'm not sure I can afford it. 

What's the cost per hour?

I am using some custom controls, and I've given them a star-trek console look just for fun. Maybe one of them is causing the memory leak. I can try using standard controls, except for the gauge.

![sysmac_20_production](https://github.com/unoplatform/uno/assets/24267381/29a42c00-cd93-4a14-98a5-90ddaf107bc5)

I've been working with WinForms for a long time so I'm really a newbie in the XAML world.

This is a old version, but you can take a look [here](https://www.youtube.com/watch?v=VvJGbypb1fc&ab_channel=mcnets)

**Reply by mcNets** [reply]
I have build a new UNO project (Windows + Skia) with just two pages and some standard buttons.

If you navigate from one page to another several times, memory is steadily increasing and even GC is called (that never happens in my app), memory is never released.

I'm not sure about that but it seems related to Microsoft.UI.Xaml.DependencyProperties libraries.

[SysmacUI.zip](https://github.com/unoplatform/uno/files/12709147/SysmacUI.zip)

**Reply by jeromelaban** [reply]
https://github.com/unoplatform/uno/pull/13958 may help for this issue.
